File: express-backend/src/models/Observation.ts
import mongoose, { Schema, type Document, type Model } from 'mongoose';

export interface ObservationDocument extends Document {
  patient_id: mongoose.Types.ObjectId;
  observation_code: string;
  value_numeric?: number;
  value_text?: string;
  unit?: string;
  effective_at?: Date;
  created_at: Date;
  neo4j_synced_at?: Date | null;
}

const ObservationSchema = new Schema<ObservationDocument>(
  {
    patient_id: { type: Schema.Types.ObjectId, ref: 'Patient', required: true },
    observation_code: { type: String, required: true },
    value_numeric: { type: Number },
    value_text: { type: String },
    unit: { type: String },
    effective_at: { type: Date },
    created_at: { type: Date, default: () => new Date() },
    neo4j_synced_at: { type: Date, default: null },
  },
  {
    collection: 'observations',
  },
);

export const Observation: Model<ObservationDocument> =
  mongoose.models.Observation ||
  mongoose.model<ObservationDocument>('Observation', ObservationSchema);

File: backend/app/schemas/observation.py
"""Observation (lab/vital value) document schema."""
from datetime import datetime, timezone
from typing import Optional

from pydantic import BaseModel, Field

from .objectid import PyObjectId


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


class ObservationDoc(BaseModel):
    """Mongo document shape for observations collection. Links to patient_id."""

    id: Optional[PyObjectId] = Field(None, alias="_id")
    patient_id: PyObjectId
    observation_code: str
    value_numeric: Optional[float] = None
    value_text: Optional[str] = None
    unit: Optional[str] = None
    effective_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=_utc_now)
    neo4j_synced_at: Optional[datetime] = None

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}
